src/parameter.jl

```diff
@@ -0,0 +1,165 @@
+    return Factor(f, T)
+end
+
+function Parameter(p::Parameter)
+    @match p begin
+        PiUnit(pu, _) => Parameter(Val(:PiUnit), pu)
+        Factor(f, _) => Parameter(Val(:Factor), f)
+    end
+end
+
+Base.copy(p::Parameter) = @match p begin
+    PiUnit(pu, _) => Parameter(Val(:PiUnit), pu)
+    Factor(f, _) => Parameter(Val(:Factor), f)
+end
+
+function Base.show(io::IO, p::Parameter)
+    @match p begin
+        PiUnit(pu, _) && if pu isa Number
+        end => print(io, "$(pu)⋅π")
+        PiUnit(pu, pu_type) && if !(pu isa Number)
+        end => print(io, "PiUnit(($(pu))::$(pu_type))")
+        Factor(f, _) => print(io, "$(f)")
+    end
+end
+
+function eqeq(p1, p2)
+    @match (p1, p2) begin
+        (PiUnit(pu1, _), PiUnit(pu2, _)) => pu1 == pu2
+        (Factor(f1, _), Factor(f2, _)) => f1 == f2
+        (PiUnit(pu1, _), Factor(f2, _)) => exp(im * pu1 * π) == f2
+        (Factor(f1, _), PiUnit(pu2, _)) => f1 == exp(im * pu2 * π)
+        (PiUnit(_...), _) => eqeq(p1, Parameter(Val(:PiUnit), p2))
+        (Factor(f1, _), ::Number) => f1 == p2
+        _ => error(
+            "Invalid input '$(p1)' of type $(typeof(p1)) and '$(p2)' of type $(typeof(p2)) for ADT: ==",
+        )
+    end
+end
+
+Base.:(==)(p1::Parameter, p2::Parameter) = eqeq(p1, p2)
+Base.:(==)(p1::Parameter, p2::Number) = eqeq(p1, p2)
+Base.:(==)(p1::Number, p2::Parameter) = eqeq(p2, p1)
+
+# following the same convention in Phase.jl implementation
+# comparison have inconsistent, we are comparing phases to numbers
+# if cause trouble, will change
+Base.isless(p1::Parameter, p2::Number) = @match p1 begin
+    PiUnit(_...) => p1.pu isa Number && p1.pu < p2
+    _ => p1.f < p2
+end
+
+Base.convert(::Type{Parameter}, p::T) where {T} = @match p begin
+    ("PiUnit", pu) => Parameter(Val(:PiUnit), pu)
+    ("Factor", f) => Parameter(Val(:Factor), f)
+    PiUnit(_...) => p
+    Factor(_...) => p
+    _ => error("Invalid input '$(p)' of type $(typeof(p)) for ADT: Parameter")
+end
+
+Base.zero(p::Parameter) = @match p begin
+    PiUnit(_...) => Parameter(Val(:PiUnit), zero(Float64))
+    Factor(_...) => Parameter(Val(:Factor), one(ComplexF64))
+end
+
+Base.one(p::Parameter) = @match p begin
+    PiUnit(_...) => Parameter(Val(:PiUnit), 1.0)
+    Factor(_...) => Parameter(Val(:Factor), exp(im * 1.0 * π))
+end
+
+Base.zero(::Type{Parameter}) = Parameter(Val(:PiUnit), 0.0)
+
+Base.one(::Type{Parameter}) = Parameter(Val(:PiUnit), 1.0)
+
+Base.iseven(p::Parameter) = @match p begin
+    PiUnit(_...) => (p.pu isa Number) && (-1)^p.pu > 0
+    Factor(_...) => p.f ≈ 1.0 # making meaning the same, upto machine precision
+end
+
+function add_param(p1, p2)
+    @match (p1, p2) begin
+        (PiUnit(pu1, _), PiUnit(pu2, _)) && if pu1 isa Number && pu2 isa Number
+        end => Parameter(Val(:PiUnit), pu1 + pu2)
+        (PiUnit(pu1, _), PiUnit(pu2, _)) && if !(pu1 isa Number) || !(pu2 isa Number)
+        end => Parameter(Val(:PiUnit), Expr(:call, :+, pu1, pu2))
```

Comment:
Here we should include the number type of the expr. See https://github.com/QuantumBFS/ZXCalculus.jl/blob/bbc19ab5ec0f1eb0b09ffa7a6e2976f60b625fbc/src/phase.jl#L30C35-L30C35

Reply:
Thank you very much! I corrected the `Expr` part. For the pattern matched above that, I declared the constructor of `Parameter` as a parametric method. The type of `pu` should be automatically matched like in the previous implementation [link](https://github.com/QuantumBFS/ZXCalculus.jl/blob/bbc19ab5ec0f1eb0b09ffa7a6e2976f60b625fbc/src/phase.jl#L27C31-L27C31) The test passed alright, I will merge the changes for now.